snuba/snuba_migrations/querylog/0006_sorting_key_change.py

```diff
@@ -0,0 +1,156 @@
+from snuba.migrations import migration, operations
+
+TABLE_NAME = "querylog_local"
+TABLE_NAME_NEW = "querylog_local_new"
+TABLE_NAME_OLD = "querylog_local_old"
+
+
+def update_querylog_table(clickhouse: ClickhousePool, database: str) -> None:
+    ((curr_create_table_statement,),) = clickhouse.execute(
+        f"SHOW CREATE TABLE {database}.{TABLE_NAME}"
+    ).results
+
+    new_sorting_key = "dataset, referrer, toStartOfDay(timestamp), request_id"
+    new_sampling_key = "request_id TTL timestamp + toIntervalDay(30)"
+
+    ((curr_sampling_key, curr_sorting_key),) = clickhouse.execute(
+        f"SELECT sampling_key, sorting_key FROM system.tables WHERE name = '{TABLE_NAME}' AND database = '{database}'"
+    ).results
+
+    new_create_table_statement = curr_create_table_statement.replace(
+        TABLE_NAME, TABLE_NAME_NEW
+    )
+
+    # Switch the sorting key
+    if curr_sorting_key != new_sorting_key:
+        assert new_create_table_statement.count(curr_sorting_key) == 1
+        new_create_table_statement = new_create_table_statement.replace(
+            curr_sorting_key, new_sorting_key
+        )
+
+    # Switch the sample by
+    if curr_sampling_key != new_sampling_key:
+        assert new_create_table_statement.count("SAMPLE BY " + curr_sampling_key) == 1
+        new_create_table_statement = new_create_table_statement.replace(
+            "SAMPLE BY " + curr_sampling_key, "SAMPLE BY " + new_sampling_key
+        )
+
+    # Update the timestamp column
+    # Clickhouse 20 does not support altering a column in the primary key so we need to do it here
+    new_timestamp_type = "`timestamp` DateTime CODEC(T64, ZSTD(1))"
+    assert new_create_table_statement.count("`timestamp` DateTime") == 1
+    new_create_table_statement = new_create_table_statement.replace(
+        "`timestamp` DateTime", new_timestamp_type
+    )
+    assert new_timestamp_type in new_create_table_statement
+
+    # Create the new table
+    clickhouse.execute(new_create_table_statement)
+
+    # Copy the data over
+    [(row_count,)] = clickhouse.execute(f"SELECT count() FROM {TABLE_NAME}").results
+    batch_size = 100000
+    batch_count = math.ceil(row_count / batch_size)
+
+    orderby = "toStartOfDay(finish_ts), project_id, event_id"
+
+    for i in range(batch_count):
+        skip = batch_size * i
+        clickhouse.execute(
+            f"""
+            INSERT INTO {TABLE_NAME_NEW}
+            SELECT * FROM {TABLE_NAME}
+            ORDER BY {orderby}
+            LIMIT {batch_size}
+            OFFSET {skip};
+            """
+        )
+
+    clickhouse.execute(f"RENAME TABLE {TABLE_NAME} TO {TABLE_NAME_OLD};")
+
+    clickhouse.execute(f"RENAME TABLE {TABLE_NAME_NEW} TO {TABLE_NAME};")
+
+    # Ensure each table has the same number of rows before deleting the old one
+    assert (
+        clickhouse.execute(f"SELECT COUNT() FROM {TABLE_NAME};").results
+        == clickhouse.execute(f"SELECT COUNT() FROM {TABLE_NAME_OLD};").results
+    )
```

Comment:
Is it possible for newer writes to happen while this loop is running https://github.com/getsentry/snuba/pull/3611/files#diff-8832387c64f114e4caa833a673d0d8b6cfb00cb5c12f77a40df3cf00f741ba08R58-R75? If yes, this assert can fail.

Reply:
Yes that is correct. Since this is a data migration you would have take the table inserts offline first. I'm not sure if there's an easy way around that.

Reply:
Got it. I would suggest you remove this assert to remove the ambiguity.